R/case-variables.R

```diff
@@ -0,0 +1,154 @@
+#' the case list.
+#' 
+#' There are two ways to specify cases, but you must pick only one (note these 
+#' two will produce the same case variable):
+#' 
+#' 1. When you just want to specify conditions, you can use named conditions:
+#' `makeCaseVaraible(case1=ds$v1 == 1, case2=ds$v2 == 2, name="new case")`
+#' 
+#' 1. You can also use the `cases` argument, which is useful when you want to 
+#' prespecify ids, or set numeric_values:
+#' `makeCaseVaraible(cases=list(list(expression=ds$v1 == 1, name="case1"), list(expression=ds$v2 == 2, name="case2")), name="new case")`
+#' 
+#' @param ... a sequence of named expressions to use as cases as well as other 
+#' properties to pass about the case variable (i.e. alias, description)
+#' @param cases a list of lists with each case condition to use each must 
+#' include at least a `name` and an `expression` element.
+#' @param else_case a single list that has at least a `name` to serve as the 
+#' case when no others match (if not specified Crunch will use the system 
+#' default "No data")
+#' @param name a character to use as the name of the case variable to create
+#' @export
+makeCaseVariable <- function (..., cases, else_case, name) {
+    dots <- list(...)
+    is_expr <- function (x) inherits(x, "CrunchLogicalExpr")
+    exprs <- Filter(is_expr, dots)
+    if (!missing(cases) & length(exprs) > 0) {
+        halt("can't have case conditions both in ... as well as in the cases ",
+             "argument, please use one or the other.")
+    }
+    if (missing(cases) & length(exprs) == 0) {
+        halt("must supply case conditions in either ... or the cases ",
+             "argument, please use one or the other.")
+    }
+    if (missing(cases) & length(exprs) > 0) {
+        # save any member of dots that isn't an expression for use later.
+        dots <- Filter(Negate(is_expr), dots)
+        cases <- mapply(function (expr, name) list(name=name, expression=expr),
+                        expr=exprs, name=names(exprs),
+                        SIMPLIFY = FALSE, USE.NAMES = FALSE)
+    }
+    
+    # make ids if the cases don't have ids.
+    need_ids <- vapply(cases, function (x) is.null(x$id), logical(1))
+    supplied_ids <- vapply(cases[!need_ids], function (x) x$id, integer(1))
+    new_ids <- seq_along(cases) # generate too many ids
+    new_ids <- setdiff(new_ids, supplied_ids) # discard ids where there is overlap
+    cases[need_ids] <- lapply(seq_along(cases[need_ids]), function(i) {
+        cases[need_ids][[i]]$id <- new_ids[i]
+        return(cases[need_ids][[i]])
+    })
+    
+    # check there are no duplicate ids
+    all_ids <- vapply(cases, function (x) x$id, integer(1))
+    if (anyDuplicated(all_ids) > 0) {
+        halt("there are duplicate ids provided: ", serialPaste(all_ids))
+    }
+    
+    cases <- lapply(cases, ensureValidCases)
+
+    if (!missing(else_case)) {
+        # assign API default values if not given
+        if(is.null(else_case$id)) {
```

Comment:
This series of `if`s could be done more concisely with `modifyList` (in `utils`, already imported elsewhere)